fix(core): properly handle app stabilization with defer blocks

alan-agius4 opened the pull request:
Previously, the app was marked as stable prematurely. For more details, see https://github.com/angular/angular/issues/61038#issuecomment-2837917180
    
Closes: #61038

alan-agius4 committed:
test: add integration test for defer with input on SSR with zones

Add an integration test to verify defer block behavior with input handling during server-side rendering (SSR).

alan-agius4 committed:
test: add integration test for platform-server with zoneless

Add an integration test to verify SSR with zoneless

alan-agius4 committed:
test: disable platform-server tests that do not work with zoneless

A number of cases currently do not work with zoneless.

devversion commented on integration/platform-server-zoneless/e2e/src/http-transferstate-lazy-on-init-spec.ts:
```diff
@@ -0,0 +1,45 @@
+import {bootstrapClientApp, navigateTo, verifyNoBrowserErrors} from './util';
+
+// TODO: this does not work with zoneless
+xdescribe('Http TransferState Lazy On Init', () => {
```
are there tracking issues for these TODOs or are you addressing them soon?

alan-agius4 replied:
I am looking at this right now

devversion commented on integration/platform-server-zoneless/projects/standalone/src/app/http-transferstate-lazy/http-transfer-state.component.ts:
```diff
@@ -0,0 +1,38 @@
+  `,
+  providers: [HttpClient],
+})
+export class TransferStateComponent implements OnInit {
```
I'm not super familiar with what is being tested here; just making sure it's intentional? I does look like those are just general tests for SSR and zoneless?

atscott commented on packages/core/src/pending_tasks.ts:
```diff
@@ -64,9 +64,17 @@ export class PendingTasksInternal implements OnDestroy {
+
+    // Recheck in a microtask to prevent a race condition caused by the gap between
+    // task removal and task creation.
+    queueMicrotask(() => {
```
This is an extremely dangerous change, one that I’ve already attempted and abandoned. 

ZoneJS stability already queues a microtask before removing the task: https://github.com/angular/angular/blob/244594694335cb37340c7e5576dd55af80c68cf8/packages/core/src/change_detection/scheduling/ng_zone_scheduling.ts#L297 this would introduce a second one. 

Stability is already deferred in the version of this api that’s already public. Internal APIs sometimes use the synchronous one to maintain previous behavior, especially as it relates to NgZone, which is very sensitive to and aware of every microtask. 

Instead of changing this, I would recommend migrating from the internal api to the public pending task service where appropriate.

alan-agius4 committed:
fix(core): properly handle app stabilization with defer blocks

Previously, the app was marked as stable prematurely. For more details, see https://github.com/angular/angular/issues/61038#issuecomment-2837917180

Closes: #61038

thePunderWoman reviewed (approved):
LGTM! Thanks for digging into this from the tooling side. It's certainly a fun one.

jsaguet-betclic commented:
Will the fix be included in the v19 branch as well?

alan-agius4 committed:
fixup! fix(core): properly handle app stabilization with defer blocks

alan-agius4 commented:
> Will the fix be included in the v19 branch as well?

Yes, there is a separate PR for that https://github.com/angular/angular/pull/61056

alan-agius4 commented:
**caretaker note:** the lint is the usual license check failure

alan-agius4 committed:
fixup! fix(core): properly handle app stabilization with defer blocks

AndrewKushnir commented:
This PR was merged into the repository by commit 193bd7c54a32a20c53927bfda6c9343cf2de12d2.

The changes were merged into the following branches: main, 20.0.x
